### PR TITLE
DDF-1133: Added resource URI conversion to open search endpoint.

### PR DESCRIPTION
**Pull request by lessarderic**
Please take a look and let me know if the approach makes sense, i.e., adding a proxy in front of the CatalogFramework that converts the resource URI before calling the transform on the Catalog.

I thought this was the simplest, most reusable and least invasive approach but I'm obviously open to comments and suggestions :-)

Once I get approval, I'll go ahead and refactor the existing CSW implementation to use the same approach and add all the proper unit tests.

[Review on Reviewable](https://reviewable.io/reviews/codice/ddf-catalog/64)

### CODE REVIEW

File: opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        transform(response);
+        return catalog.transform(response, transformerId, requestProperties);
+    }
+
+    public UpdateResponse update(UpdateRequest updateRequest) throws IngestException,
+        SourceUnavailableException {
+        return catalog.update(updateRequest);
+    }
+
+    private void transform(SourceResponse queryResponse) {
```
Comment (mrmateo) [inline opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java]: Hey @lessarderic , I like the idea but I'm a little uneasy about having the proxy. Couple quick things that make me uneasy about this approach --  1) the amount of code that just calls the internal framework and 2) the fact that doing this takes away the endpoint using the osgi framework directly for the framework. This means that different endpoints will be using different implementations of frameworks and I think that violates some of the concepts of having the central framework.

I do like the add-on feature though and think it's definitely useful. Is this something you can see every endpoint using? I know you mentioned CSW, if so it might be worthwhile to look at creating a new type of plugin (pretransformplugin?) and have the catalogframework just call that plugin before doing a transform. This would be acceptable from my POV as we're doing a minor version change (2.7.0) which allows for added functionality in the api.

Hopefully this helps, let me know if you have any questions!

Comment (lessarderic) [inline opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java]: Hi @mrmateo,

We do want to re-use that code for other endpoints, so my plan was to move that proxy class to the ddf.catalog.impl package and have any endpoint that needs that URI conversion use it. I did have a discussion with @menousek and @pklinef about doing this using a plug-in as you suggested. The general feeling was that this is still a fairly specific use case and adding yet another plug-in would increase complexity with little value added at this point.

If it's okay with you, I'd like to wait and see what other people think and go from there.

Thanks a lot for the feedback!

Eric

File: opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/OpenSearchEndpoint.java [hunk 2]
```diff
@@ -190,10 +181,8 @@ public Response processQuery(@QueryParam(PHRASE)
                 Set<String> siteSet = new HashSet<String>(Arrays.asList(StringUtils
                         .stripAll(sources.split(","))));
 
-                // This code block is for backward compatibility to support
-                // src=local.
-                // Since local is a magic work, not in any specification, we
-                // need to
+                // This code block is for backward compatibility to support src=local.
+                // Since local is a magic work, not in any specification, weneed to
```
Comment (kcwire) [inline opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/OpenSearchEndpoint.java]: lost a space in the comment "weneed"

File: opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java (hunk 1 above)
Comment (kcwire) [inline opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java]: I disagree that this is a specific use case. The idea is simple, if a client calls your service they should be provided a metacard that is useable on their system. If we return resource-uri's with our internal form (ie content:asdfasdfadsf) then they will not be able to do anything useful with that without building additional logic into their client.  If we provide them "https://host/services/catalog/source/asdfasdfasdf?transform=resource" thats something very useable from a client.

I agree with @mrmateo about abstracting the catalog framework, I'm not crazy about that. I would rather see a preTransform plugin or just this business logic baked in directly to the endpoints.

Comment (lessarderic) [inline opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java]: I guess the question really boils down to the following: do we want this URI conversion to be (a) baked in specific endpoints or (b) do we want it to be optional/configurable for any endpoint?

If the answer is a) then in my mind the current solution using the proxy achieves this and also allows the logic to be re-used between endpoints that need the behavior to be "baked-in". Also, since it's internal, it doesn't require any analysis in terms of external and future impact.

If we feel that this conversion should be optional and we see a need for other pre-transform logic down the road, then I'm all for option b). However, if we go with that option, I feel that we should spend some time defining exactly what the plug-in interface should look like and what those plug-ins should be allowed to do. We may also want to consider writing a configurable plug-in that allows conversion of any attributes present in the response, not only the resource URI, and consider whether we want to keep using ActionProviders/Actions to do so.

In other words, I have nothing against the plug-in approach. I just think that before we expose a new plug-in and use it for this use case, we should spend some time thinking through all the use cases to make sure we expose the right thing and don't open the door to something we didn't intend.

Thoughts?

Eric

File: opensearch/catalog-opensearch-endpoint/pom.xml [hunk 3]
```diff
@@ -142,6 +147,8 @@
               org.objectweb.asm.tree,
               org.objectweb.asm.tree.analysis,
               org.objectweb.asm.util,
+              ddf.catalog.impl,
```
Comment (pklinef) [inline opensearch/catalog-opensearch-endpoint/pom.xml]: The ddf.catalog.*.impl packages should be embedded and not imported.  Looks like the current imports are incorrect too.

File: opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+import ddf.catalog.source.SourceUnavailableException;
+import ddf.catalog.source.UnsupportedQueryException;
+import ddf.catalog.transform.CatalogTransformerException;
+
+/**
+ * {@link CatalogFramework} proxy class used to transform the URIs found in a {@link SourceResponse}
+ * {@link Metacard}s using the injected {@link ActionProvider} before transformation by the
+ * {@link CatalogFramework}.
+ */
+public class ResourceUriTransformerCatalogFrameworkProxy implements ddf.catalog.CatalogFramework {
```
Comment (pklinef) [inline opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java]: Why was the ddf.catalog package not imported?

File: opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java (hunk 1 above)
Comment (pklinef) [inline opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java]: I think that all endpoints will want this behavior but not all catalog framework clients are guaranteed to be endpoints.  If this is baked into the catalog framework or added as a plugin, then all catalog framework clients will only have access to the external resource URL.  There might be cases where that is not desired.  For example, layered federation or a scheduled job that deletes metacards and any associated local content.

Assuming everyone agrees that we should not force this on all catalog framework clients, then we either need some kind of utility class with this logic for endpoints to reuse or we can use this proxy in a dependency injection framework.  I think I like the proxy pattern over more utility classes.

Comment (mrmateo) [inline opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java]: That's a good point @pklinef about clients not always being endpoints.

During the brainstorm process did you guys look at adding the external url as a different attribute in the metacard (instead of overwriting the current uri)? That would leverage the extensibility of the metacard interface, not add code dependencies in the endpoints (i.e. require using the proxy for future endpoints), and maintain the internal uri that may be of use in future components. It would be a very minor change to the current transformers and allows the possibility of using the internal uri as a fallback for backwards compatibility.

Comment (kcwire) [inline opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java]: +1 @mrmateo - It seems like we have finally run into a scenario where we truly have multiple definitions for a single attribute.  I think a new attribute is the correct direction moving forward.

Comment (lessarderic) [inline opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java]: I do like the idea of having two distinct attributes and we did briefly talk about it.

There's a couple of things we need to consider if we go down that path though:

1. Is our goal with this ticket to convert the URI for the clients so we don't need to change the clients' code itself? If so, then using a separate attribute will not achieve this as client code will have to be changed to use the new attribute instead of the resource-uri one.

2. There's kind of a tricky naming issue here. I assume the new attribute would be name something like resource-url. Since URLs are also URIs, it might cause some confusion from a client perspective. What I'm trying to say here is that using resource-urn and resource-url would be better names for these attributes, but that would mean deprecating resource-uri and I'm not sure if we want/can/should do that.

@pklinef, do you remember if there were any other reasons why we thought adding a new attribute would be a problem?

Comment (mrmateo) [inline opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java]: Some comments to your thoughts:
1. The client code doesn't have to be updated, only the transformer code does. Additionally, maybe formats support multiple resource urls so this could be just adding a new url to the format. With the proxy implementation you're still having to 'change' the clients by modifying their dependency injection to include the proxy.
2. I would prefer having a talk about something as simple as deciding a new attribute name rather than updating all current endpoints and requiring all future endpoints to embed / use a proxy catalogframework in order to get something as simple as an external url.

Comment (pklinef) [inline opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java]: We only brainstormed replacing the current resource uri value but not adding a new one.  I think adding an attribute that contains a URL generated by the system resource download action provider makes sense.  Most transformers and endpoints will just work with that change.  We just won't be able to add a new getter for that attribute on the metacard interface but I think that is acceptable.

Comment (lessarderic) [inline opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java]: Looks like we have a winner then :-)

I'll close this PR and submit a new one with the code change needed to create the new attribute after the query has been executed so this new attribute is always there.

Eric

Comment (pklinef) [inline opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java]: You can leave this PR open and just force push different commits to your branch if you want to.

Comment (jvettraino) [inline opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java]: I have been following this thread and wanted to ask a couple of questions and present my thoughts.

It seems the information that you want (the download URL) is exactly what the Action Framework is designed to do.  So shouldn't the transformers use the Action Framework as designed, and then set the values accordingly in their specific output format?  In the case of the XML Transformer - perhaps it makes sense to have elements related to actions.  It seems like the proxy approach and even the additional attribute approach are mixing metacard concepts and muddying things up, rather than having the clear separate between Metacard Attributes and Actions that can be done on a Metacard.

I would think you would want to make the XML Transformer (and any transformers) Action aware.   Likewise any clients using that response format should be aware of the format of the client.

I don't think the Resource URL (which is retrieved by the Resource Action), should have any special handling that other Actions (thumbnail URL, Metadata URL, etc.) don't have.

Just a thought :)

Comment (kcwire) [inline opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java]: @jvettraino I was thinking the same thing originally, and then I had a change of heart.  I think what it comes down to is from an endpoint (ie Opensearch, CSW, ECDR, etc) we want to provide the "download-link" in the metadata, vs requiring an external client, implementing a standard, to know how to construct a URL the an ActionProvider could just provide in the original response.

What do you mean by making the transformers "Action aware"?

Comment (jvettraino) [inline opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java]: @kcwire I am in total agreement with the desire for an endpoint to provide the "download-link" however I don't think the endpoints ability to expose links should be dictated by the Metacard attributes.  Rather the endpoint should be taking the Metacard and transforming it into a format that is suitable for it.  Meaning it is in the transformation process where both first-class (Metacard Attributes) and derived (URLS from Actions - e.g. the download-link) should be used to provide the endpoint with a format that meets their needs (this transformation can be manual inside the endpoint or in a *Transformer implementation).

So for the original problem in this ticket, I would actually chalk this up as a bug/limitation in the XML Transformer, and that is where I would look to make the change.  If we are saying that the XML view of a Metacard should include the download-link, then shouldn't it happen in the code that actually transforms the Metacard to XML?  

Isn't this download-link more relevant to the external metadata representation (format that endpoints expose via Metacard Transformation) rather than the internal Metacard representation (Metacard attributes)?

Comment (lessarderic) [inline opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java]: Ok, now I get it. You're point @jvettraino is that the new URL should be added to the output of the transformation based on the current resource-uri instead of being forced inside the metacard as a new attribute.

I've been looking at the CatalogFrameworkImpl class and I think I kind of agree with this now. Adding a new attribute to the metacard outside of the ingest path doesn't feel quite right.

Our original plan was to *replace* the resource-uri with the download link inside the transformer but that didn't seem right either as it takes away something that might actually be useful.

Let me see how difficult it would be to add the attribute when the transform is done instead of adding it to the metacard.

Comment (jvettraino) [inline opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java]: @lessarderic if you go that route, one thing you may want to think about is, since the XML Transformer is a generic XML view of a Metacard, does it make sense to have ALL actions generically be represented in the XML

Comment (menousek) [inline opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java]: Putting actions into the XML representation is risky, simply because they are somewhat transient and tied to the DDF node that generates the XML.  Normally there wouldn't be hostnames tied back to DDF in the XML representation.

I think there is a strong case to have a resource-url attribute that clients could use directly.  We have been running into those a lot lately (I opened a Jira for this based on comments from Mario Savard).

That resource-uri attribute could be populated a number of ways.  It seems that we are coming across a need for some kind of "enricher" pattern.  Metacard in, "better" Metacard out.  That is also basically the same as the Pre/Post Ingest/Query Plugins so we could use that as a place to trigger these if we don't want to introduce new interfaces (and I'd rather not).

Comment (jvettraino) [inline opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java]: If I could go back and ask, what is the key problem we are trying to solve? Is it endpoints getting a URL where they could retrieve the Resource?

If that is it, then I think the pattern is already in place.
 * Endpoints use transformers to get data in the format they need
 * Action Framework returns the URL for how to retrieve a Resource

@menousek I understand your concern about jamming actions in the XML representation of a Metacard since the XML representation is supposed to be generic, although I am not sure I agree with the hostname thing since if you included resource-url as an attribute it would have a hostname (maybe I am missing something) - and I also don't see it as a bad thing to include this info (whats the downside)?.  Of course the actions are a snapshot in time, of what can be done with a metacard, rather than persistent attributes, so you would want to represent them accordingly.

The MetacardEnricher I like :) , but I go back to the fact we already have the Resource URL from the Action Framework, and it is really not something that would be derived from an enricher.  The URL seems to be an Action, not an enrichment (enrichment seems more tied to analytics, or adding additional info).

**Comment by lessarderic** [comment]
Updated the pull request based on the comments provided. @pklinef, @jvettraino, @mrmateo and @kcwire, please let me know if you're good with this new solution.

Will add the CSW changes as well as the unit tests to this PR later on today.

File: core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessor.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+ * License, or any later version.
+ * 
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ * 
+ **/
+package ddf.catalog.impl;
```
Comment (kcwire) [inline core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessor.java]: Interfaces should not go in an "impl" package.  Should be ddf.catalog.blah instead.

File: core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessorImpl.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+ * License, or any later version.
+ * 
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ * 
+ **/
+package ddf.catalog.impl;
```
Comment (kcwire) [inline core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessorImpl.java]: this would then be ddf.catalog.blah.impl :)

File: core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessor.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+import ddf.catalog.operation.QueryResponse;
+
+/**
+ * Interface implemented by classes that can be used to perform post-
+ * 
+ * @author elessard
```
Comment (kcwire) [inline core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessor.java]: We don't do author tags. only git will give you credit.

File: core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessor.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ * 
+ **/
+package ddf.catalog.impl;
+
+import ddf.catalog.operation.QueryResponse;
+
+/**
+ * Interface implemented by classes that can be used to perform post-
```
Comment (kcwire) [inline core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessor.java]: Javadoc appears incomplete

File: core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessorImpl.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    /**
+     * Converts the resource URIs found in the {@link SourceResponse} {@link Metacard}s using the
+     * {@link ActionProvider} injected in the constructor before delegating the call to the proxied
+     * {@link CatalogFramework} object.
+     */
+    @Override
+    public void processResponse(QueryResponse queryResponse) {
+        // TODO Auto-generated method stub
+
+        for (Result result : queryResponse.getResults()) {
```
Comment (kcwire) [inline core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessorImpl.java]: Check is the resourceActionProvider is not null first instead of every loop.

File: core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessorImpl.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+ * {@link CatalogFramework}.
+ */
+public class QueryResponsePostProcessorImpl implements QueryResponsePostProcessor {
+    // PreCatalogTransformPlugin {
+    private static final Logger LOGGER = LoggerFactory
+            .getLogger(QueryResponsePostProcessorImpl.class);
+
+    private ActionProvider resourceActionProvider;
+
+    public QueryResponsePostProcessorImpl() {
```
Comment (kcwire) [inline core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessorImpl.java]: Is this ever used? Seems like this is kind of useless without the resourceActionProvider

File: opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/OpenSearchEndpoint.java [hunk 11]
```diff
@@ -14,6 +14,32 @@
  **/
```
Comment (kcwire) [inline opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/OpenSearchEndpoint.java]: So there should be no changes to the OpensearchEndpoint right? You could just checkout these changes so we aren't just reorganizing the imports.

File: core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessor.java (hunk 5 above)
Comment (lessarderic) [inline core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessor.java]: Good point. I'll move it in the ddf.catalog package. Since this is not a real API interface (yet), I'll keep it in the catalog-core-standardframework bundle if you're ok with that.

File: core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessorImpl.java (hunk 6 above)
Comment (lessarderic) [inline core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessorImpl.java]: Any suggestion on the sub-package name? blah doesn't look too good ;-)

File: core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessor.java (hunk 7 above)
Comment (lessarderic) [inline core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessor.java]: Sorry. Eclipse added it for me. Need to change my config.

File: core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessor.java (hunk 8 above)
Comment (lessarderic) [inline core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessor.java]: Beefed up the comment.

File: core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessorImpl.java (hunk 10 above)
Comment (lessarderic) [inline core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessorImpl.java]: Removed.

File: core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessorImpl.java (hunk 9 above)
Comment (lessarderic) [inline core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessorImpl.java]: Good catch, thanks!

File: core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessorImpl.java (hunk 10 above)
Comment (lessarderic) [inline core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessorImpl.java]: Removed.

File: core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessorImpl.java (hunk 6 above)
Comment (kcwire) [inline core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessorImpl.java]: operation, plugin, or a new package. I don't have a string opinion other than I don't think this should be in just ddf.catalog

Comment (lessarderic) [inline core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/QueryResponsePostProcessorImpl.java]: Since I created this class to avoid adding any more code to the CatalogFrameworkImpl class and make it easier to unit test, I left it in the same package for now, got rid of the interface and renamed the class (see today's commit). The CatalogFrameworkImpl class would need some major refactoring and unit testing so I thought we could always revisit this when we come back to it. Let me know what you think.

File: core/catalog-core-api-impl/src/main/java/ddf/catalog/data/impl/BasicTypes.java [hunk 12]
```diff
@@ -306,6 +306,9 @@ public AttributeFormat getAttributeFormat() {
                 true /* stored */, false /* tokenized */, false /* multivalued */, XML_TYPE));
         descriptors.add(new AttributeDescriptorImpl(Metacard.RESOURCE_URI, true /* indexed */,
                 true /* stored */, false /* tokenized */, false /* multivalued */, STRING_TYPE));
+        descriptors.add(new AttributeDescriptorImpl(Metacard.RESOURCE_DOWNLOAD_URL,
+                true /* indexed */, true /* stored */, false /* tokenized */,
+                false /* multivalued */, STRING_TYPE));
```
Comment (jvettraino) [inline core/catalog-core-api-impl/src/main/java/ddf/catalog/data/impl/BasicTypes.java]: This confuses me.  If you want to include a dynamic action value (aka the download URL) as a metacard attribute, the value shouldn't indexed, or stored, in fact I don't think it should be included in the Metacard BasicTypes definition at all.  

Assuming it doesn't make sense to be included in the BasicTypes definition, then that backsup that point that the design for including the download URL as a Metacard attribute is flawed and that the download-url should not really be a Metacard attribute, but rather an action that can be derived from a Metacard (which it currently is today, so I am still confused by the desire to include it as an attribute as I think it leads to more confusion and more complex logic). 

Also wouldn't you then have to update the endpoint code to specifically use this value?  I didn't see that in this PR.

Comment (lessarderic) [inline core/catalog-core-api-impl/src/main/java/ddf/catalog/data/impl/BasicTypes.java]: Good catch. This was added by mistake in a previous commit. Undoing the change.

Regarding your other point, we've debated this quite a bit and came to the conclusion that it was simpler to always include it in the metacard on the way out. That way, all endpoints will always return it and clients will not have to explicitly convert the resource-uri or have to deal use actions.

This could also have been achieved using a post-query plugin but we felt that there wasn't a good use case to create a plug-in that would always be installed by default and shouldn't be removed.

Hope this helps. @pklinef and @kcwire can provide more information on this if you'd like.

**Comment by shaundmorris** [comment]
@lessarderic @pklinef @kcwire @jvettraino do we still need more dialog on this PR?

**Comment by jvettraino** [comment]
I strongly standby my comments that this is muddying the design quite a bit, and could be a source for confusion down the road, in that you are now placing an Action (which already has a framework devoted to it), as a first class attribute of Metacard (thus confusing Metacard attributes with Metacard Actions).  So essentially this is creating a one-off solution for a specific action/case (which likely will not be the only time this pops up).

But I understand the majority rules :)

**Comment by menousek** [comment]
@lessarderic @pklinef @kcwire I would like to talk about @jvettraino's comments today especially before this is potentially merged/closed.

**Comment by kcwire** [comment]
@menousek please set something up ;)
